src/api_manager/api_manager_impl.cc

```diff
@@ -291,35 +282,41 @@ ApiManager::ApiBasepathRewriteAction ApiManagerImpl::ReWriteURL(
     return ApiBasepathRewriteAction::NONE;
   }
 
-  std::size_t path_end_pos = url.find_first_of('?');
-  std::string path =
-      (path_end_pos == std::string::npos) ? url : url.substr(0, path_end_pos);
-
-  bool append_slash = false;
-  if (path.length() == 0 || path.at(path.length() - 1) != '/') {
-    path.append("/");
-    append_slash = true;
+  std::size_t path_length = url.find_first_of('?');
+  if (path_length == std::string::npos) {
+    path_length = url.length();
   }
 
   auto base = server_config->api_service_config().base_path();
 
-  if (CheckBasePathMatch(path, base) == false) {
-    if (server_config->api_service_config().base_path_hard_match()) {
-      return ApiBasepathRewriteAction::REJECT;
+  if (base.length() > path_length) {
+    return server_config->api_service_config().base_path_hard_match()
+               ? ApiBasepathRewriteAction::REJECT
+               : ApiBasepathRewriteAction::NONE;
+  } else if (base.length() == path_length) {
+    if (url.compare(0, path_length, base) == 0) {
+      destination_url->assign("/");
+      destination_url->append(url.substr(path_length));
+      return ApiBasepathRewriteAction::REWRITE;
     } else {
       return ApiBasepathRewriteAction::NONE;
```

Comment:
do we need to check hard or soft for this return?

It seems that will be easier to have a match function first,  call that function first, then check hard or soft in the out side layer.

Reply:
Fixed the bug and added test case.